lib/NormalModule.js

```diff
@@ -353,47 +360,35 @@ class NormalModule extends Module {
 		};
 		const loaderContext = {
 			version: 2,
-			getOptions: (loaderName, schema) => {
-				const loader = loaderContext.loaders[loaderContext.loaderIndex];
-				let options = {};
-
-				if (loader.options && typeof loader.options === "object") {
-					({ options } = loader);
-				} else if (loader.query) {
-					let { query } = loader;
-
-					if (query.substr(0, 1) !== "?") {
-						throw new WebpackError(
-							"A valid query string should begin with '?'"
-						);
-					}
+			getOptions: schema => {
+				const loader = this.getCurrentLoader(loaderContext);
 
-					query = query.substr(1);
+				let { options } = loader;
 
-					// Allow to use `?foo=bar` in `options`
-					if (query.substr(0, 1) === "?") {
-						query = query.substr(1);
-					}
-
-					if (query.substr(0, 1) === "{" && query.substr(-1) === "}") {
+				if (typeof options === "string") {
+					if (options.substr(0, 1) === "{" && options.substr(-1) === "}") {
 						try {
-							options = parseJson(query);
+							options = parseJson(options);
 						} catch (e) {
-							throw new WebpackError(`Cannot parse query string: ${e.message}`);
+							throw new Error(`Cannot parse string options: ${e.message}`);
 						}
 					} else {
-						options = querystring.parse(query);
+						options = querystring.parse(options, "&", "=", {
+							maxKeys: 0
+						});
 					}
 				}
 
-				if (!schema) {
-					return options;
+				if (options === null || options === undefined) {
+					options = {};
 				}
 
-				validateOptions(schema, options, {
-					name: loaderName || "Unknown Loader",
-					baseDataPath: "options"
-				});
+				if (schema) {
+					validateOptions(schema, options, {
+						name: getCurrentLoaderName(),
```

Comment:
What if i need to setup name for a loader? `schema-utils` prefer the name argument under name in schema

Reply:
Yep, fix that. Now it uses `schema.title`, but falls back to `Loader` and `options` (instead of the default `Object` and `configuration`).

Maybe `fallbackName` and `fallbackBaseDataPath` in the schema-utils API would be nice.